File: 2021/fsharp/Day01.fs
namespace AdventOfCode.FSharp.Y2021

// Day 1: Sonar Sweep
module Day01 =
    open AdventOfCode.FSharp.Util

    let increasing input =
        input
        |> Seq.pairwise
        |> Seq.filter (fun (a, b) -> (b > a))
        |> Seq.length

    let part1 input =
        input
        |> splitLine
        |> Seq.map int
        |> increasing
        |> string

    let part2 input =
        input
        |> splitLine
        |> Seq.map int
        |> Seq.windowed 3
        |> Seq.map Seq.sum
        |> increasing
        |> string
        
    let run input (output: int -> string -> unit) =
        let numbers =
            input
            |> splitLine
            |> Array.map int

        let part1 =
            numbers
            |> increasing
            |> string
        output 1 part1

        let part2 =
            numbers
            |> Seq.windowed 3
            |> Seq.map Seq.sum
            |> increasing
            |> string
        output 2 part2

File: 2021/csharp/Day01.cs
namespace AdventOfCode.CSharp.Y2021;

public static class ArrayExtension
{
    public static IEnumerable<ArraySegment<T>> Window<T>(this T[] input, int window)
    {
        ArraySegment<T> root = input;
        for (var i = 0; i < input.Length; ++i)
        {
            if (i + window <= input.Length)
                yield return root.Slice(i, window);
        }
    }
}

public static class Day01
{
    public static void Run(byte[] input, Action<int, string> output)
    {
        var Input = Encoding.UTF8
            .GetString(input)
            .Split('\n', StringSplitOptions.RemoveEmptyEntries);

        int Part1()
        {
            var numbers = Input.Select(int.Parse);

            return Increasing(numbers);
        }

        int Part2()
        {
            var windows = Input
                .Select(int.Parse)
                .ToArray()
                .Window(3)
                .Select(s => s.Sum());

            return Increasing(windows);
        }

        output(1, Part1().ToString());
        output(2, Part2().ToString());
    }

    static int Increasing(IEnumerable<int> numbers)
    {
        var last = int.MaxValue;
        var total = 0;
        foreach (var n in numbers)
        {
            if (n > last) total += 1;
            last = n;
        }
        return total;
    }
}